ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Storage/LedgerDB/V2/LedgerSeq.hs

```diff
@@ -0,0 +1,479 @@
+{-# LANGUAGE DataKinds                  #-}
+{-# LANGUAGE DeriveAnyClass             #-}
+{-# LANGUAGE DeriveGeneric              #-}
+{-# LANGUAGE DerivingVia                #-}
+{-# LANGUAGE FlexibleContexts           #-}
+{-# LANGUAGE FlexibleInstances          #-}
+{-# LANGUAGE GADTs                      #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE MultiParamTypeClasses      #-}
+{-# LANGUAGE QuantifiedConstraints      #-}
+{-# LANGUAGE RankNTypes                 #-}
+{-# LANGUAGE ScopedTypeVariables        #-}
+{-# LANGUAGE StandaloneDeriving         #-}
+{-# LANGUAGE TypeOperators              #-}
+{-# LANGUAGE UndecidableInstances       #-}
+
+-- | The data structure that holds the cached ledger states.
+module Ouroboros.Consensus.Storage.LedgerDB.V2.LedgerSeq (
+    -- * LedgerHandles
+    LedgerTablesHandle (..)
+    -- * The ledger seq
+  , LedgerSeq (..)
+  , LedgerSeq'
+  , StateRef (..)
+  , closeLedgerSeq
+  , empty
+  , empty'
+    -- * Apply Blocks
+  , applyBlock
+  , applyThenPush
+  , extend
+  , prune
+  , pruneToImmTipOnly
+    -- * Queries
+  , anchor
+  , current
+  , currentHandle
+  , getPastLedgerAt
+  , immutableTipSlot
+  , isSaturated
+  , maxRollback
+  , rollback
+  , rollbackN
+  , rollbackToAnchor
+  , rollbackToPoint
+  , snapshots
+  , tip
+  , volatileStatesBimap
+  ) where
+
+import           Data.Function (on)
+import           Data.Word
+import           GHC.Generics
+import           NoThunks.Class
+import           Ouroboros.Consensus.Block
+import           Ouroboros.Consensus.Config.SecurityParam
+import           Ouroboros.Consensus.Ledger.Abstract
+import           Ouroboros.Consensus.Ledger.Extended
+import           Ouroboros.Consensus.Ledger.Tables.Utils
+import           Ouroboros.Consensus.Storage.LedgerDB.Config
+import           Ouroboros.Consensus.Util.IOLike
+import           Ouroboros.Network.AnchoredSeq hiding (anchor, last, map,
+                     rollback)
+import qualified Ouroboros.Network.AnchoredSeq as AS hiding (map)
+import           Prelude hiding (read)
+import           System.FS.API
+import           System.FS.API.Types
+
+{-------------------------------------------------------------------------------
+  LedgerTablesHandles
+-------------------------------------------------------------------------------}
+
+data LedgerTablesHandle m l = IOLike m => LedgerTablesHandle {
+    close       :: !(m ())
+  , duplicate   :: !(m (LedgerTablesHandle m l))
+  , read        :: !(LedgerTables l KeysMK -> m (LedgerTables l ValuesMK))
+  , readAll     :: !(m (LedgerTables l ValuesMK))
+  , write       :: !(LedgerTables l DiffMK -> m ())
+  , writeToDisk :: !(SomeHasFS m -> FsPath -> m ())
```

Comment:
Make sure that this is compatible with LSM-tree style snapshots (see [the docs](https://input-output-hk.github.io/lsm-tree/lsm-tree/Database-LSMTree-Normal.html#v:snapshot)). LSM-tree style snapshots don't allow snapshotting into arbitrary directories. The snapshot directory is fixed, but names for the snapshots can be picked.

So, maybe it should be something like `String -> m ()`

Reply:
Done